runtime/src/snapshot_utils.rs

```diff
@@ -1136,54 +1143,58 @@ pub(crate) fn parse_incremental_snapshot_archive_filename(
     })
 }
 
-/// Get a list of the full snapshot archives in a directory
-pub fn get_full_snapshot_archives<P>(snapshot_archives_dir: P) -> Vec<FullSnapshotArchiveInfo>
-where
-    P: AsRef<Path>,
-{
-    match fs::read_dir(&snapshot_archives_dir) {
-        Err(err) => {
+/// Walk down the snapshot archive directory recursively to collect snapshot archive file info
+fn get_snapshot_archives<T>(dir: &Path, cb: fn(PathBuf) -> Result<T>, output: &mut Vec<T>) {
```

Comment:
I still feel like this fn can be simplified, and made more safe. Some things:

- At line 1161, `p.ends_with(SNAPSHOT_ARCHIVE_DOWNLOAD_DIR)`, I don't like the "ends_with" check. Since that means there could be other directories that incorrectly get read here, i.e. `backup_remote`.

- We get the snapshot archives dir from the snapshot config, and we have a fn `build_snapshot_archives_remote_dir()` to get the remote directory. I think `build_snapshot_archives_remote_dir()` should be used, and not reimplement that logic in this function.

- I think overall it'd be simpler to get all the snapshots in the snapshot archives dir, and then get all the snapshots in the remote dir. Walking the directory recursively seems unnecessary, and potentially an issue if somehow there are `remote/remote/remote/remote/../remote/` directories with the same or duplicated snapshots. 

- I'd like to see this fn return the snapshot archive infos instead of using a `&mut` parameter.

Reply:
yup. Fixed.